nlp_primitives/openai/embeddings.py

```diff
@@ -0,0 +1,186 @@
+            -0.007940744049847126,
+            0.007481361739337444,
+            ...
+            0.009351702407002449,
+            -0.016065239906311035
+          ],
+          [
+            -0.001055666827596724,
+            0.01066350657492876,
+            ...
+            -0.024650879204273224,
+            0.009666346944868565
+          ],
+          [
+            nan,
+            nan,
+            ...
+            nan,
+            nan
+          ],
+        ]
+    """
+
+    name = "openai_embeddings"
+    input_types = [ColumnSchema(logical_type=NaturalLanguage)]
+    return_type = ColumnSchema(logical_type=Double, semantic_tags={"numeric"})
+
+    def __init__(
+        self,
+        model: OpenAIEmbeddingModel = DEFAULT_MODEL,
+        max_tokens_per_batch: Optional[int] = None,
+    ):
+        self.model = model
+        self.number_output_features = model.output_dimensions
+        if max_tokens_per_batch is None:
+            self.max_tokens_per_batch = model.max_tokens * 10
+        else:
+            self.max_tokens_per_batch = max_tokens_per_batch
+
+    def _is_too_many_tokens(self, element, encoding) -> bool:
+        """Return whether a data element has too many tokens and should be ignored"""
+        return len(encoding.encode(element)) > self.model.max_tokens
+
+    def _create_request_batches(
+        self, series
+    ) -> List[OpenAIRequest[OpenAIEmbeddingResponse]]:
+        """Group elements of a series into batches of requests to send to OpenAI"""
+        # the encoding used by the model
+        encoding = tiktoken.get_encoding(self.model.encoding)
+
+        # a static embeddings response for an invalid element
+        invalid_request = StaticOpenAIEmbeddingRequest(
+            result=[np.nan] * self.number_output_features
+        )
+
+        # mutable variables to track the request batching process
+        # a running list of the requests to make
+        requests: List[OpenAIRequest[OpenAIEmbeddingResponse]] = []
+        # a list of elements that should be batched into the next request
+        elements_in_batch = []
+        # a running total of tokens that will be sent in the next request
+        tokens_in_batch = 0
+
+        def add_batched_request() -> int:
+            """Create a batched request from the currently staged elements and add it to the request list
+            """
+            if elements_in_batch:
+                elements_copy = list(elements_in_batch)
+                elements_in_batch.clear()
+                requests.append(
+                    OpenAIEmbeddingRequest(
+                        list_of_text=elements_copy,
+                        model=self.model,
+                        token_consumption=tokens_in_batch,
+                    )
+                )
+                return 0
+            else:
+                return tokens_in_batch
+
+        def can_fit_in_batch(tokens) -> bool:
+            return (
+                len(elements_in_batch) < 2048
```

Comment:
What is the significance of 2048?

Reply:
yeah, here's the limit in the openai client: https://github.com/openai/openai-python/blob/main/openai/embeddings_utils.py#L43